feat: Add SerdeObject support for pasta curves

CPerezz opened the pull request:
With the current serialization primitives we have here, pasta curves did not have any way to be serialized as we did not implement `SerdeObject` for it.

This PR adds this support which resolves
https://github.com/privacy-scaling-explorations/halo2curves/issues/23. The code needs to have `unsafe` chunks as we need to transmute memory in order to be able to access the private fields of the pasta-curves crate structs.

This should be revisited and re-thought once
https://github.com/privacy-scaling-explorations/halo2/issues/176 is addressed.

CPerezz commented:
Should I add unit-tests too? I do think so TBH. But wanted to know if they're maybe already done somewhere else in halo2.

han0110 commented on src/pasta/mod.rs:
```diff
@@ -25,3 +29,300 @@ impl crate::CurveAffineExt for EqAffine {
+
+        let elt = Fp::from(AccessibleFp(inner));
+        unsafe {
+            is_less_than(&inner, &std::mem::transmute_copy(&Fp::MODULUS))
```
`Fp::MODULUS` is in big-endian, should we hardcode the modulus as constant in little-endian somewhere and use that for this comparison?

ed255 commented on src/pasta/mod.rs:
```diff
@@ -25,3 +29,300 @@ impl crate::CurveAffineExt for EqAffine {
+    }
+}
+
+struct AccessibleVestaPoint {
```
This one has the same code as `AccessiblePallasPoint`, the only difference is the call to `.is_on_curve()`.  So I also wonder if the code can be deduplicated :thinking:

ed255 reviewed (commented):
The implementation looks correct to me!  I've added a few comments, mostly about:
- Reducing the number of `unsafe`
- Exploring code deduplication

I definitely think there should be unit tests here, for each type and for different values: do the serialization and deserialization and test equality.

CPerezz commented on src/pasta/mod.rs:
```diff
@@ -25,3 +29,300 @@ impl crate::CurveAffineExt for EqAffine {
+        // The transmute is the only way to get access to the internal array.
+        // Note that this is safe as long as the underlying array length is the same
+        // and the types contained are the same too.
+        let limbs: [u64; 4] = unsafe { std::mem::transmute_copy(&self) };
```
This will force us to clone when not needed in reality. THat's why I don't like it. 
I can do it if you feel like it has to. But I'd prefer avoiding and having faster serialization.

CPerezz commented:
Addressed most of the review comments.

Pending actions:
- [ ] Re-write as macros to de-duplicate code.
- [ ] Write unit tests for serialization/deserialization.
